# :tanabata_tree: JustMAQT

[![R](https://img.shields.io/badge/R-276DC3?style=for-the-badge&logo=R&logoColor=white&labelColor=101010)](https://www.r-project.org/about.html)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/mathiashole/JustMAQT?style=for-the-badge&labelColor=101010&color=white)
![GitHub](https://img.shields.io/github/license/mathiashole/GScissors?color=%23179287&style=for-the-badge&logoColor=white&labelColor=101010)

`JustMAQT` is an R script for customizing and visualizing phylogenetic trees.

## :book: Features

-   Plotting phylogenetic trees using input from a variety of file formats.
-   Customizable tree layouts (rectangular, circular, equal_angle, etc.).
-   Color-coding tree branches based on specified keywords.
-   Option to root the tree at a specified ID.
-   Clustering options 
-   Heatmaps option for continuous values or genotypes.
-   Command-line interface.

## :hammer: Usage

To use the phylogenetic tree plotting script, follow these steps:

-   Clone or download the script from the repository [Link](https://github.com/mathiashole/JustMAQT).
-   Ensure you have the required R packages: ggtree, treeio, ape, RColorBrewer, optparse, cluster, factoextra, and ggplot2.
-   Open a terminal and navigate to the folder where the script is located.
-   Run the script with the necessary arguments. Key options include the phylogenetic tree file, keywords for coloring, tree layout, and optional files for alignment, genotype, or continuous values.

## :hammer: in progress ...


## :sparkling_heart: Contributing

- :octocat: [Pull requests](https://github.com/mathiashole/JustMAQT/pulls) and :star2: stars are always welcome.
- For major changes, please open an [issue](https://github.com/mathiashole/JustMAQT/issues) first to discuss what you would like to change.
- Please make sure to update tests as appropriate.

## :mega: Contact

:bird: [Mathias](https://twitter.com/joaquinmangino)

## License
MIT &copy; [Mathias Mangino](https://github.com/mathiashole)